text/0000-add-filename-info-to-io-error.md

```diff
@@ -0,0 +1,130 @@
+
+# Summary
+
+[summary]: #summary
+
+This RFC proposes to add filename information to `std::io:Error`. By doing so, `std::io` error messages will include which file the user is trying to access and making it easier for the user to notice where their code is failing.
+
+# Motivation
+
+[motivation]: #motivation
+
+Consider a simple program which reads a file which doesn't exist:
+
+```rust
+use std::fs::File;
+
+fn main() -> std::io::Result<()> {
+    let file = File::open("hello.txt")?;
+    println!("file: {:?}", file);
+    Ok(())
+}
+```
+
+This program emits the following error message:
+
+```txt
+Error: Os { code: 2, kind: NotFound, message: "No such file or directory" }
+```
+
+This error message doesn't tell the user which file doesn't exist and doesn't give the user a full breakdown of the error and what went wrong. As I've experienced and many others have, this error doesn't help a beginner Rust programmer to understand what went wrong and how to fix it. If the compiler emitted a full error and more information such as the name of the file which doesn't exist, this would be much more helpful.
+
+# Guide-level explanation
+
+[guide-level-explanation]: #guide-level-explanation
+
+Suppose a user creates the same program as before:
+
+```rust
+use std::fs::File;
+
+fn main() -> std::io::Result<()> {
+    let file = File::open("hello.txt")?;
+    println!("file: {:?}", file);
+    Ok(())
+}
+```
+
+With this change, a much more descriptive and helpful error message will be emitted:
+
+```txt
+Error: Os { code: 2, kind: NotFound, message: "No such file or directory", file: "/Users/henryboisdequin/hello.txt" }
+```
+
+This error explains where the error is emitted and the cause of it. As you can see, this error is much more helpful. It points to where the problem is and the cause of it. Let's look at another case:
+
+```rust
+use std::fs::File;
+
+fn main() -> std::io::Result<()> {
+    let _ = File::create("/etc/protocols")?;
+
+    Ok(())
+}
+```
+
+This program returns the following error:
+
+```txt
+Error: Os { code: 13, kind: PermissionDenied, message: "Permission denied" }
+```
+
+This program is not helpful at all, not displaying file information or how to fix this error. With this change, here is the projected error message:
+
+```
+Error: Os { code: 13, kind: PermissionDenied, message: "Permission denied", file: "/etc/protocols" }
+```
+
+In this case, this change would add the file to the `Os` error message. Many crates and Rust applications write their own error messages to improve the experience while working with `std::io`. This is why we should incorporate filename information to `std::io:Error` to reduce the amount of boilerplate crates and Rust applications need to write to ensure a better `std::io` development process.
+
+# Reference-level explanation
+
+[reference-level-explanation]: #reference-level-explanation
+
+1\.
+
+- Add field `file: Option<PathBuf>` to struct `std::io::Error`
```

Comment:
Something to consider: some file operations deal with more than one path, for example [`std::fs::copy`](https://doc.rust-lang.org/std/fs/fn.copy.html) and [`std::fs::rename`](https://doc.rust-lang.org/std/fs/fn.rename.html). I'm not sure if there are any that deal with more than two paths though, in which case `(Option<PathBuf>, Option<PathBuf>)` would be sufficient.